rfc/system/5103-offloaded-literal.md

```diff
@@ -4,32 +4,77 @@
 
 ## 2 Motivation
-A [cursory search](https://discuss.flyte.org/?threads%5Bquery%5D=LIMIT_EXCEEDED) of Slack history shows a few times that this has come up before (and I remember some other instances, I think that search term just wasn't included). This is something that we've historically addressed by just increasing the size of the grpc message that's allowed, but this is an unsustainable solution.
+A [cursory search](https://discuss.flyte.org/?threads%5Bquery%5D=LIMIT_EXCEEDED) of Slack history shows a few times that this has come up before (and I remember some other instances, I think that search term just wasn't included). This is something that we've historically addressed by just increasing the size of the grpc message that's allowed, but this is an unsustainable solution and severely reduces the utility of large-fan-out map tasks.
 
 ## 3 Proposed Implementation
+We propose configuring propeller to offload large literal collections, using the following config
 
-### 3.1 Offloaded Literal IDL
-To the `Literal` [message](https://github.com/flyteorg/flyte/blob/cb6384ac6ea60f8b9421a71cfda4279f3579d3cb/flyteidl/protos/flyteidl/core/literals.proto#L95), add a new field called `starp` that will point to a location in the "metadata" bucket of the Flyte backend. The offloaded bytes should be deserialzable into a `Literal` object.
+```yaml
+type LiteralOffloadingConfig struct {
+  Enabled bool
+  // Maps flytekit SDK to minimum supported version that can handle reading offloaded literals.
+  SupportedSDKVersions map[string]string
+  // Default, 10Mbs. Determines the size of a literal at which to trigger offloading
+  MinSizeInMBForOffloading uint64
+  // Fail fast threshold
+  MaxSizeInMBForOffloading uint64
+}
+
+```
 
-Questions: How will things like metadata be handled? Should they be merged? What should be in the `value` field of the main parent Literal?
+### 3.1 Offloaded Literal IDL
+Update the `Literal` [message](https://github.com/flyteorg/flyte/blob/4a7c3c0040b1995a43939407b99ca3e87b1eb752/flyteidl/protos/flyteidl/core/literals.proto#L94-L114)
+like so
+
+```protobuf
+message Literal {
+    oneof value {
+        // A simple value.
+        Scalar scalar = 1;
+        // A collection of literals to allow nesting.
+        LiteralCollection collection = 2;
+        // A map of strings to literals.
+        LiteralMap map = 3;
+    }
+    ...
+    // ** new below this line **
+    // If this literal is offloaded, this field will contain metadata including the offload location.
+    string uri = 6;
+    // Includes information about the size of the literal.
+    uint64 size_bytes = 7;
```

Comment:
why is size important again?  I mean there's other metadata as well (etag information). The assumption here is that size is super important so we want to be able to show that without making a head call?

Reply:
general consensus seemed to be this is useful, I guess it's nice for clients who want to decide whether to pull massive datasets?